### ISSUE TITLE
Cholesky Factorization For Complex Matrix Broken

### ISSUE THREAD

**Issue by JaredCrean2** [JuliaLang/julia#14804]
The `chol()` function for complex matrices seems to be ignoring the complex part of the matrix:

``` julia
a = [6. 2 3; 2 9 5; 3 5 11]
b = a + complex(0, 10)*eye(3)
println("b = \n", b)
U = chol(b)
b2 = U'*U

println("b2 = \n", b2)  # no complex part
```

outputs:

``` julia
b = 
Complex{Float64}[6.0 + 10.0im 2.0 + 0.0im 3.0 + 0.0im
                 2.0 + 0.0im 9.0 + 10.0im 5.0 + 0.0im
                 3.0 + 0.0im 5.0 + 0.0im 11.0 + 10.0im]
b2 = 
Complex{Float64}[5.999999999999999 + 0.0im 2.0 + 0.0im 3.0 + 0.0im
                 2.0 + 0.0im 9.000000000000002 + 0.0im 5.0 + 0.0im
                 3.0 + 0.0im 5.0 + 0.0im 11.0 + 0.0im]
```

versioninfo:

```
Julia Version 0.4.3-pre+6
Commit adffe19* (2015-12-11 00:38 UTC)
Platform Info:
  System: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7 CPU       Q 820  @ 1.73GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Nehalem)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

**Comment by JaredCrean2** [comment]
Sorry, bad test case as the matrix is *not* hermitian.  However, a warning here might be preferable to silently returning the wrong answer

**Comment by tkelman** [comment]
Ignoring the imaginary component of the diagonal here seems like a bad fortran convention that we should avoid inheriting from lapack.

**Comment by andreasnoack** [comment]
I think it is safe to assume that the BLAS and LAPACK people made this choice on purpose. That is not the same as saying that we should do the same thing, but in real use cases, the diagonal will almost never have zero imaginary parts, e.g.
```jl
julia> A = randn(5,5); A = A'A;

julia> B = complex(randn(5,5), randn(5,5)); imag(diag(B'A*B))
5-element Array{Float64,1}:
  8.32667e-16
  0.0        
 -1.77636e-15
 -8.88178e-16
  0.0        
```
We are handling `B'B` specially such there it is exactly `Hermitian`, but in most real code this is not sufficient. So it is a tradeoff between getting errors/warnings all the time when in fact there are no issues weigthed against a few cases where such a warning/error catches a genuine error.

A few alternatives to status quo are

1. Remove the check in `Hermitian` for zero imaginary part in the diagonal and insist that `Hermitian` is used whenever the user wants to ignore any imaginary noise. @tkelman and I discussed this recently in https://github.com/JuliaLang/julia/pull/10672

2. Introduce a tolerance. We discussed this to the death of the PR in https://github.com/JuliaLang/julia/pull/10369

**Comment by JaredCrean2** [comment]
That's a good point, I'm not sure there is a good standard for deciding what numerical zero is independent of the application.  Having a kwarg for a tolerance that is tight be default and letting the user loosen it if needed seems like the only way to catch errors and allow all the behavior we want.  I don't think its too onerous to require the users of numerical language to set tolerances to simple things like this.

On the other hand, having complex numbers on the diagonal is a bit of a special case, so maybe just documenting it and retaining the existing behavior is good enough.

**Comment by tkelman** [comment]
BLAS and LAPACK were designed without consideration for generic programming. The implementer of every single routine that ever touches an HE or PO matrix has to take this special case into account and manually ignore complex components of diagonal elements. This is not a good convention to follow for generic programming in Julia where users will be directly manipulating the underlying `.data` storage fairly often and for a variety of purposes.

With the exception of the reflectors which are quite different from the rest of the structured matrix types, Julia's structured matrix types are generally indexing view transformations of the wrapped data. Adding this one special case where specifically for elements on the diagonal, that view transformation also involves only considering part of the scalar stored at that location, seems error prone and easy to miss for implementers of algorithms designed to work on generic array types. I think we can come up with a better solution to the tolerancing issue.

**Comment by andreasnoack** [comment]
Fixed by #16799
